File: 7. vigenere/vigenere.c
/****************************************************************************
 * vigenere.c
 *
 * Programmeren I
 * Problem set 2 (Crypto)
 * Minor Programmeren
 * Najib el Moussaoui
 *
 * Input a key, and cipher your input-text by +key (alphabetical word)
 ***************************************************************************/

#include <cs50.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <ctype.h>

// initialize function
bool convert();

int main(int argc, string argv[])
{

    // check if argument count passed in to main is 2
    if (argc != 2)
    {
        printf("Error1\n");
        return 1;
    }

    // pull phrase to use for cipher from argv[1], store in variable key
    string key = argv[1];

    // loop over argv[1] to check if input is all alphabetical
    for (int i = 0, n = strlen(key); i < n; i++)
    {
        if (!isalpha(key[i]))
        {
            printf("Error2\n");
            return 1;
        }
    }

    // prompt for plain text to cipher
    printf("plaintext: ");
    string t = get_string();

    // output ciphertext
    printf("ciphertext: ");
    for (int i = 0, n = strlen(t), j = 0; i < n; i++)
    {
        bool converted = convert(t[i], key[j]);
        if (converted)
        {
            j = (j + 1) % strlen(key);
        }
    }

    printf("\n");
}

// function that takes plain text and key to convert to ciphertext
bool convert(letter, key)
{
    int keyalpha;
    int c;

    // check if letter in key is lower of upper, substract by ASCII value
    if (isupper(key))
    {
        keyalpha = key - 'A';
    }
    else if (islower(key))
    {
        keyalpha = key - 'a';
    }

    // convert to ciphertext
    if (isupper(letter))
    {
        c = ((letter - 'A') + keyalpha) % 26;
        printf("%c", c + 'A');
        return true;
    }
    else if (islower(letter))
    {
        c = ((letter - 'a') + keyalpha) % 26;
        printf("%c", c + 'a');
        return true;
    }
    else
    {
        printf("%c", letter);
        return false;
    }
}

File: 7. vigenere/vigenere.py
############################################################################
# vigenere.py
#
# Minor Programming
# Programming II
# Problem Set 6: Sentimental
# Najib el Moussaoui
#
# Encrypt a message with a special alphabetical key.
###########################################################################

import cs50 as cs
import sys

# ensure proper usage
if len(sys.argv) != 2:
    print("Use two command-line arguments bro")
    exit(1)

# extract the secret key from command-line (sssht)
key = sys.argv[1]

# check if cipherkey is all alphabetical
if key.isalpha() == False:
    print("Only alphabetical characters pls")
    exit(2)

# prompt user for plaintext to cipher
plain = cs.get_string("plaintext: ")

# counters
index = 0
j = 0


def convert(letter):
    """Returns an index between 0-25 for every letter."""
    if letter.islower():
        return ord(letter) - ord('a')
    elif letter.isupper():
        return ord(letter) - ord('A')

print("ciphertext: ", end="")

# loop over plaintext, and shift with key's index, with wraparound.
for i in range(len(plain)):
    if plain[i].isalpha():
        index = (convert(plain[i]) + convert(key[j % len(key)])) % 26
        j += 1
    if plain[i].isupper():
        print(chr(index + ord('A')), end="")
    elif plain[i].islower():
        print(chr(index + ord('a')), end="")
    else:
        print(plain[i], end="")

# Aloha and au revoir
print()



